src/esp/core/Buffer.h

```diff
@@ -26,12 +27,12 @@ enum class DataType {
 
 class Buffer {
  public:
-  explicit Buffer(){};
+  explicit Buffer() {}
   explicit Buffer(const std::vector<size_t> shape, const DataType dataType) {
```

Comment:
Oh, the `Buffer` is still here? Somehow I was under the impression that it got removed/replaced already :)

Reply:
IIRC on Slack we discussed using the [StridedArrayView](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1StridedArrayView.html) instead, since it's a superset of the features provided by this one, it's fully tested, has a conversion to any numpy array and is generally better integrated to Magnum. But I don't have the full context either, maybe @angelxuanchang and @msbaines could comment here? :)

In case it's not clear -- I'm not saying this PR should be taking any action on this, just got surprised it wasn't handled already elsewhere.

Reply:
> StridedArrayView requires the shape of the data to be known at compile time

Not sure if we are talking about the same thing -- it only needs to know the dimensionality, *not* sizes. (There's StaticArrayView, which has compile-time sizes, but that's a different thing.) Dimensionality is, in this particular case, 2. When potentially used for audio (as @angelxuanchang mentioned on Slack) it would be 1, for volumetric data it would be 3, but that's it. Based on my experience, having dimensionality templated makes C++ code much easier to reason about, with no need for allocting an unbounded `std::vector` holding the size, with no need to check for its size at runtime, similarly as having a distinct type for two- and three-component vectors is beneficial. And I don't think this makes bindings harder -- for things with different dimension count you'd need a completely different set of APIs anyway, so why not a type as well.

Reply:
I see, that's a valid concern. :+1: 

A different idea, then -- what about always using the max dimension count, so 3D? Then there's no templating not inheritance hierarchy needed; if the observation is one-dimensional, the second and third dimension is always one; if the observation is two-dimensional, the third dimension is always one. Furthermore, on the C++ side, converting between different dimensions is just a matter of calling [slice&lt;N&gt;()](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1StridedArrayView.html#a66c3b7ae7c67cad7896b68202ebff081), so if some code would really expect just one/two dimensions, the conversion is an one-liner (of course with a check that nothing got lost when dropping the extra dimensions):

```cpp
Containers::StridedArrayView3D<float> observations;
Containers::StridedArrayView1D<float> samples = observations.slice<1>();
```

"Inflating" to more dimensions works similarly, setting size of the newly added dimensions to `1`.

Reply:
I *knew* you'd point that out :stuck_out_tongue_winking_eye:  

The container can be "type-erased" (`char`, e.g.) and then cast to a concrete type at the point of use. That's what [Image::pixels()](https://doc.magnum.graphics/magnum/classMagnum_1_1Image.html#a2fa5831183fcd118cb3e4040a258e895) does already, for example.